### PR TITLE
FW-1437: Publish Categories on Create and Update

### PR DESCRIPTION
**Pull request by davidlacho**

### CODE REVIEW

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java [hunk 1]
```diff
@@ -20,74 +20,31 @@
...
 import org.nuxeo.ecm.automation.core.collectors.DocumentModelCollector;
 import org.nuxeo.ecm.core.api.ConcurrentUpdateException;
-import org.nuxeo.ecm.core.api.CoreSession;
 import org.nuxeo.ecm.core.api.DocumentModel;
-import org.nuxeo.ecm.core.api.IdRef;
-import org.nuxeo.ecm.core.api.PathRef;
+import org.nuxeo.runtime.api.Framework;
 
 @Operation(id = UpdateCategory.ID, category = Constants.CAT_DOCUMENT, label = "Update Category",
     description =
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java:35]: @davidlacho we should update the description to be more specific.

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java [hunk 2]
```diff
@@ -20,74 +20,31 @@
...
 
-  @Context
-  protected CoreSession session;
+  public static final String ID = "Document.UpdateCategory";
 
   @Param(name = "properties", required = false)
   protected Map<String, String> properties;
 
   @OperationMethod(collector = DocumentModelCollector.class)
   public DocumentModel run(DocumentModel doc) throws ConcurrentUpdateException {
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java:47]: Is there a reason this throws ConcurrentUpdateException? does the service throw that?

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      throw new InvalidCategoryException("Document type must be FVCategory.");
+    }
+
+    if (properties.size() > 0) {
+      for (Map.Entry<String, String> entry : properties.entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+
+        if (key.equals("ecm:parentRef")) {
+          session.saveDocument(doc);
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java:33]: Why are we saving the document here?

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+          doc.setPropertyValue(key, value);
+        }
+      }
+    }
+    session.saveDocument(doc);
+
+    return publishDocumentIfDialectPublished(session, doc);
+  }
+
+  private DocumentModel publishDocumentIfDialectPublished(CoreSession session, DocumentModel doc) {
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java:65]: consider naming this publishCategory, and either adding "if dialect published" as a comment (or consider it implied in line 70)

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    if (!doc.getType().equals("FVCategory")) {
+      throw new InvalidCategoryException("Document type must be FVCategory.");
+    }
+
+    if (properties.size() > 0) {
+      for (Map.Entry<String, String> entry : properties.entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+
+        if (key.equals("ecm:parentRef")) {
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java:32]: comment would be helpful - what is parentRef? why is it a special case?

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          session.saveDocument(doc);
+
+          if (value.contains("/Categories")) {
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java:35]: Not clear to me what we are checking for here / why

File: FirstVoicesOperations/src/main/java/services/AbstractFirstVoicesOperationsService.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package services;
+
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+
+/**
+ * @author david
+ */
+public abstract class AbstractFirstVoicesOperationsService {
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/java/services/AbstractFirstVoicesOperationsService.java:9]: This is going to be very convenient!

File: FirstVoicesOperations/src/main/resources/META-INF/MANIFEST.MF [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+Bundle-SymbolicName: FirstVoicesOperations
+Bundle-Name: FirstVoicesOperations
+Require-Bundle: org.nuxeo.runtime
+Bundle-Version: 1.0.qualifier
+Bundle-ClassPath: .
+Bundle-ActivationPolicy: lazy
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-ManifestVersion: 2
+Bundle-Vendor: Nuxeo
+Nuxeo-component: OSGI-INF/dialect/categories/categories-operations.xml,OSGI-INF/dialect/categories/categories-services.xml
```
Comment (dy2288) [inline FirstVoicesOperations/src/main/resources/META-INF/MANIFEST.MF:11]: Small thing for readability - consider new lines for this (remember MANIFEST.MF needs a new line at the end, and spaces between the xml files - see other MANIFEST.MFs)

File: FirstVoicesPublisher/src/main/java/ca/firstvoices/publisher/listeners/AssignOriginListener.java [hunk 9]
```diff
@@ -80,5 +80,9 @@ public void handleEvent(Event event) {
       saveOrigin(doc, relatedAudio);
       saveOrigin(doc, relatedVideos);
     }
+
+    if (doc.getType().equals("FVCategory")) {
```
Comment (dy2288) [inline FirstVoicesPublisher/src/main/java/ca/firstvoices/publisher/listeners/AssignOriginListener.java:84]: What does this do?

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java (hunk 2 above)
Comment (davidlacho) [reply FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/operations/UpdateCategory.java:47]: Left over from old code, will remove.

File: FirstVoicesPublisher/src/main/java/ca/firstvoices/publisher/listeners/AssignOriginListener.java (hunk 9 above)
Comment (davidlacho) [reply FirstVoicesPublisher/src/main/java/ca/firstvoices/publisher/listeners/AssignOriginListener.java:84]: Might not be the best place for it, but we're handling publishing on updating a category, but we're not handling publishing on a new category. I'd expect that if we are handling new categories in their own operations in the future we would consider moving this.

File: FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            value = session.getDocument(new PathRef(value)).getId();
+          }
+
+          DocumentModel parent = session.getDocument(doc.getParentRef());
+
+          if (!value.equals(parent.getId())) {
```
[the remark this answers is not in the document]
Comment (davidlacho) [reply FirstVoicesOperations/src/main/java/ca/firstvoices/dialect/categories/services/CategoryServiceImpl.java:44]: Comment added

File: FirstVoicesMaintenance/src/main/java/ca/firstvoices/maintenance/dialect/categories/Constants.java [hunk 11]
```diff
@@ -22,6 +22,10 @@
 
 public class Constants {
 
+  private Constants() {
+    throw new IllegalStateException("Utility class");
```
Comment (davidlacho) [inline FirstVoicesMaintenance/src/main/java/ca/firstvoices/maintenance/dialect/categories/Constants.java:26]: Utility classes, which are collections of static members, are not meant to be instantiated. Even abstract utility classes, which can be extended, should not have public constructors. Java adds an implicit public constructor to every class which does not define at least one explicitly. Hence, at least one non-public constructor should be defined.

File: .github/workflows/build.yml [hunk 12]
```diff
@@ -4,6 +4,9 @@ on:
   pull_request:
     branches:
       - master
+  push:
+    branches:
+      master
```
Comment (davidlacho) [inline .github/workflows/build.yml:9]: Adds build to master so that we can get sonar analysis on this PR when merged

**Pull request merged**